File: javascript/src/openai/openai.ts
import OpenAI from "openai";
import type {
  LanguageModel,
  LanguageModelMetadata,
} from "../models/language-model.js";
import type {
  AssistantMessage,
  AudioContainer,
  AudioEncoding,
  AudioPart,
  ContentDelta,
  LanguageModelInput,
  ModelResponse,
  ModelTokensDetail,
  ModelUsage,
  PartialModelResponse,
  TextPart,
  Tool,
  ToolCallPart,
} from "../schemas/index.js";
import { convertAudioPartsToTextParts } from "../utils/message.utils.js";
import type { InternalContentDelta } from "../utils/stream.utils.js";
import { ContentDeltaAccumulator } from "../utils/stream.utils.js";
import { calculateCost } from "../utils/usage.utils.js";
import { OpenAIRefusedError } from "./errors.js";
import type {
  OpenAIModelOptions,
  OpenAIPatchedCompletionTokenDetails,
  OpenAIPatchedPromptTokensDetails,
} from "./types.js";

const OPENAI_AUDIO_SAMPLE_RATE = 24_000;
const OPENAI_AUDIO_CHANNELS = 1;

export type OpenAILanguageModelInput = LanguageModelInput & {
  extra?: Partial<OpenAI.Chat.Completions.ChatCompletionCreateParams>;
};

export class OpenAIModel implements LanguageModel {
  provider: string;
  modelId: string;
  metadata?: LanguageModelMetadata;

  private openai: OpenAI;

  constructor(
    public options: OpenAIModelOptions,
    metadata?: LanguageModelMetadata,
  ) {
    this.provider = "openai";
    this.modelId = options.modelId;
    if (metadata) this.metadata = metadata;

    this.openai = new OpenAI({
      baseURL: options.baseURL,
      apiKey: options.apiKey,
    });
  }

  async generate(input: OpenAILanguageModelInput): Promise<ModelResponse> {
    const openaiInput = {
      ...convertToOpenAIParams(input, this.options),
      stream: false,
    } satisfies OpenAI.Chat.ChatCompletionCreateParams;

    const response = await this.openai.chat.completions.create(openaiInput);

    if (!response.choices[0]) {
      throw new Error("no choices in response");
    }

    const choice = response.choices[0];

    if (choice.message.refusal) {
      throw new OpenAIRefusedError(choice.message.refusal);
    }

    const usage = response.usage ? mapOpenAIUsage(response.usage) : undefined;

    return {
      content: mapOpenAIMessage(choice.message, openaiInput).content,
      ...(usage && { usage }),
      ...(this.metadata?.pricing &&
        usage && {
          cost: calculateCost(usage, this.metadata.pricing),
        }),
    };
  }

  async *stream(
    input: OpenAILanguageModelInput,
  ): AsyncGenerator<PartialModelResponse, ModelResponse> {
    const openaiInput = {
      ...convertToOpenAIParams(input, this.options),
      stream: true,
      stream_options: {
        include_usage: true,
      },
    } satisfies OpenAI.Chat.ChatCompletionCreateParams;

    const stream = await this.openai.chat.completions.create(openaiInput);

    let usage: ModelUsage | undefined;
    let refusal = "";
    const accumulator = new ContentDeltaAccumulator();

    for await (const chunk of stream) {
      const choice = (
        chunk.choices as
          | OpenAI.Chat.Completions.ChatCompletionChunk.Choice[]
          | undefined
      )?.[0];
      const completion = choice;

      if (completion?.delta.refusal) {
        refusal += completion.delta.refusal;
      }

      if (completion?.delta) {
        const incomingContentDeltas = mapOpenAIDelta(
          completion.delta,
          openaiInput,
          accumulator.deltas,
        );

        accumulator.addChunks(incomingContentDeltas);

        for (const delta of incomingContentDeltas) {
          yield { delta };
        }
      }

      if (chunk.usage) {
        usage = mapOpenAIUsage(chunk.usage);
      }
    }

    if (refusal) {
      throw new OpenAIRefusedError(refusal);
    }

    return {
      content: accumulator.computeContent(),
      ...(usage && { usage }),
      ...(this.metadata?.pricing &&
        usage && {
          cost: calculateCost(usage, this.metadata.pricing),
        }),
    };
  }
}

export function convertToOpenAIParams(
  input: LanguageModelInput,
  options: OpenAIModelOptions,
): OpenAI.Chat.ChatCompletionCreateParams {
  const tool_choice = convertToOpenAIToolChoice(input.toolChoice);
  const response_format = convertToOpenAIResponseFormat(
    input.responseFormat,
    options,
  );
  const samplingParams = convertToOpenAISamplingParams(input);

  return {
    model: options.modelId,
    messages: convertToOpenAIMessages(input, options),
    ...(input.tools && {
      tools: convertToOpenAITools(input.tools, options),
    }),
    ...(tool_choice && { tool_choice }),
    ...samplingParams,
    ...(response_format && {
      response_format,
    }),
    ...(input.modalities && {
      modalities: input.modalities,
    }),
    ...input.extra,
  };
}

export function convertToOpenAIMessages(
  input: Pick<LanguageModelInput, "messages" | "systemPrompt">,
  options: OpenAIModelOptions,
): OpenAI.Chat.ChatCompletionMessageParam[] {
  const openaiMessages: OpenAI.Chat.ChatCompletionMessageParam[] = [];
  let messages = input.messages;

  if (options.convertAudioPartsToTextParts) {
    messages = messages.map(convertAudioPartsToTextParts);
  }

  if (input.systemPrompt) {
    openaiMessages.push({
      role: "system",
      content: input.systemPrompt,
    });
  }

  for (const message of messages) {
    switch (message.role) {
      case "assistant": {
        const openaiMessageParam: OpenAI.Chat.ChatCompletionAssistantMessageParam =
          {
            role: "assistant",
            content: null,
          };
        message.content.forEach((part) => {
          switch (part.type) {
            case "text": {
              openaiMessageParam.content = [
                ...(openaiMessageParam.content || []),
                {
                  type: "text",
                  text: part.text,
                },
              ] as Array<OpenAI.Chat.ChatCompletionContentPartText>;
              break;
            }
            case "tool-call": {
              openaiMessageParam.tool_calls = [
                ...(openaiMessageParam.tool_calls || []),
                {
                  type: "function",
                  id: part.toolCallId,
                  function: {
                    name: part.toolName,
                    arguments: JSON.stringify(part.args),
                  },
                },
              ];
              break;
            }
            case "audio": {
              throw new Error(
                "Audio parts are not supported in assistant messages. Use options.convertAudioPartsToTextParts" +
                  " to convert audio parts to text parts.",
              );
            }
            default: {
              throw new Error(
                `Unsupported message part type: ${(part as { type: string }).type}`,
              );
            }
          }
        });
        openaiMessages.push(openaiMessageParam);
        break;
      }
      case "tool": {
        message.content.forEach((toolResult) => {
          openaiMessages.push({
            role: "tool",
            content: JSON.stringify(toolResult.result),
            tool_call_id: toolResult.toolCallId,
          });
        });
        break;
      }
      case "user": {
        const contentParts = message.content;
        openaiMessages.push({
          role: "user",
          content: contentParts.map(
            (part): OpenAI.Chat.ChatCompletionContentPart => {
              switch (part.type) {
                case "text": {
                  return {
                    type: "text",
                    text: part.text,
                  };
                }
                case "image": {
                  return {
                    type: "image_url",
                    image_url: {
                      url: `data:${part.mimeType};base64,${part.imageData}`,
                    },
                  };
                }
                case "audio": {
                  return {
                    type: "input_audio",
                    input_audio: {
                      // this as assertion is not correct, but we will rely on OpenAI to throw an error
                      format: (convertToOpenAIAudioFormat(
                        part.container,
                        part.encoding,
                      ) ||
                        "wav") as OpenAI.ChatCompletionContentPartInputAudio.InputAudio["format"],
                      data: part.audioData,
                    },
                  };
                }
                default: {
                  throw new Error(
                    `Unsupported message part type: ${(part as { type: string }).type}`,
                  );
                }
              }
            },
          ),
        });
        break;
      }
      default: {
        const exhaustiveCheck: never = message;
        throw new Error(
          `Unsupported message role: ${(exhaustiveCheck as { role: string }).role}`,
        );
      }
    }
  }

  return openaiMessages;
}

export function convertToOpenAISamplingParams(
  input: Partial<LanguageModelInput>,
) {
  return {
    ...(typeof input.maxTokens === "number" && { max_tokens: input.maxTokens }),
    ...(typeof input.temperature === "number" && {
      temperature: input.temperature,
    }),
    ...(typeof input.topP === "number" && { top_p: input.topP }),
    ...(typeof input.presencePenalty === "number" && {
      presence_penalty: input.presencePenalty,
    }),
    ...(typeof input.frequencyPenalty === "number" && {
      frequency_penalty: input.frequencyPenalty,
    }),
    ...(typeof input.seed === "number" && { seed: input.seed }),
  } satisfies Partial<OpenAI.Chat.ChatCompletionCreateParams>;
}

export function convertToOpenAIToolChoice(
  toolChoice: LanguageModelInput["toolChoice"],
): OpenAI.Chat.Completions.ChatCompletionToolChoiceOption | undefined {
  if (!toolChoice) return undefined;

  switch (toolChoice.type) {
    case "tool":
      return {
        type: "function",
        function: {
          name: toolChoice.toolName,
        },
      };
    default:
      // 1-1 mapping with openai tool choice
      return toolChoice.type;
  }
}

export function convertToOpenAITools(
  tools: Tool[],
  options: OpenAIModelOptions,
): OpenAI.Chat.Completions.ChatCompletionTool[] {
  return tools.map(
    (tool): OpenAI.Chat.Completions.ChatCompletionTool => ({
      type: "function",
      function: {
        ...(options.structuredOutputs && {
          strict: true,
        }),
        name: tool.name,
        description: tool.description,
        ...(tool.parameters && {
          parameters: tool.parameters,
        }),
      },
    }),
  );
}
// based on:
// - OpenAI.Chat.ChatCompletionAudioParam["format"]
// - https://platform.openai.com/docs/guides/speech-to-text
// - https://platform.openai.com/docs/api-reference/realtime-client-events/session/update
// - https://platform.openai.com/docs/guides/text-to-speech#supported-output-formats
// The returned type might not match the current specification on the OpenAI API
// but we still return in case future support is added
type PossibleOpenAIAudioFormat =
  | "wav"
  | "mp3"
  | "flac"
  | "opus"
  | "aac"
  | "pcm16"
  | "g711_ulaw"
  | "g711_alaw";

/**
 * Maps the audio container and encoding to the OpenAI audio format.
 * Loosely maps based on either container or encoding when defined.
 */
export function convertToOpenAIAudioFormat(
  container: AudioContainer | undefined,
  encoding: AudioEncoding | undefined,
): PossibleOpenAIAudioFormat | undefined {
  if (!container && !encoding) return undefined;

  const encodingMapping: {
    [key in AudioEncoding]: PossibleOpenAIAudioFormat;
  } = {
    flac: "flac",
    mp3: "mp3",
    opus: "opus",
    linear16: "pcm16",
    mulaw: "g711_ulaw",
    alaw: "g711_alaw",
    aac: "aac",
  };
  const containerMapping: {
    [key in AudioContainer]: PossibleOpenAIAudioFormat | undefined;
  } = {
    wav: "wav",
    ogg: "opus",
    flac: "flac",
    webm: undefined,
  };
  if (container) {
    return containerMapping[container];
  }
  if (encoding) {
    return encodingMapping[encoding];
  }
  return undefined;
}

export function convertToOpenAIResponseFormat(
  responseFormat: LanguageModelInput["responseFormat"],
  options: Pick<OpenAIModelOptions, "structuredOutputs">,
):
  | OpenAI.Chat.Completions.ChatCompletionCreateParams["response_format"]
  | undefined {
  if (!responseFormat) return undefined;

  switch (responseFormat.type) {
    case "json":
      if (options.structuredOutputs && responseFormat.schema) {
        const schemaTitle = responseFormat.schema["title"] as
          | string
          | undefined;
        const schemaDescription = responseFormat.schema["description"] as
          | string
          | undefined;
        return {
          type: "json_schema",
          json_schema: {
            strict: true,
            name: schemaTitle || "response",
            ...(schemaDescription && { description: schemaDescription }),
            schema: responseFormat.schema,
          },
        };
      }
      return { type: "json_object" };
    case "text":
      return { type: "text" };
    default: {
      const exhaustiveCheck: never = responseFormat;
      throw new Error(
        `Unsupported response format: ${(exhaustiveCheck as { type: string }).type}`,
      );
    }
  }
}

export function mapOpenAIAudioFormat(format: PossibleOpenAIAudioFormat): {
  container?: AudioContainer;
  encoding?: AudioEncoding;
} {
  switch (format) {
    case "wav":
      return { container: "wav", encoding: "linear16" };
    case "mp3":
      return { encoding: "mp3" };
    case "flac":
      return { container: "flac", encoding: "flac" };
    case "opus":
      return { encoding: "opus" };
    case "pcm16":
      return { encoding: "linear16" };
    case "aac":
      return { encoding: "aac" };
    case "g711_ulaw":
      return { encoding: "mulaw" };
    case "g711_alaw":
      return { encoding: "alaw" };
    default: {
      const exhaustiveCheck: never = format;
      throw new Error(`Unsupported audio format: ${exhaustiveCheck as string}`);
    }
  }
}

export function mapOpenAIMessage(
  message: OpenAI.Chat.Completions.ChatCompletionMessage,
  options?: Partial<OpenAI.Chat.ChatCompletionCreateParams>,
): AssistantMessage {
  return {
    role: "assistant",
    content: [
      ...(message.content
        ? [{ type: "text", text: message.content } as TextPart]
        : []),
      ...(message.audio
        ? [
            {
              type: "audio",
              ...mapOpenAIAudioFormat(options?.audio?.format || "pcm16"),
              ...(options?.audio?.format === "pcm16" && {
                sampleRate: OPENAI_AUDIO_SAMPLE_RATE,
                channels: OPENAI_AUDIO_CHANNELS,
              }),
              audioData: message.audio.data,
              transcript: message.audio.transcript,
            } satisfies AudioPart,
          ]
        : []),
      // tool call and content of openai are separate, so we define
      // an order here where the text content comes first, followed by
      // tool calls.
      ...(message.tool_calls
        ? message.tool_calls.map(
            (toolCall): ToolCallPart => ({
              type: "tool-call",
              toolCallId: toolCall.id,
              toolName: toolCall.function.name,
              args: JSON.parse(toolCall.function.arguments) as {
                [key: string]: unknown;
              },
            }),
          )
        : []),
    ],
  };
}

export function mapOpenAIDelta(
  delta: OpenAI.Chat.Completions.ChatCompletionChunk.Choice.Delta & {
    audio?: { id?: string; data?: string; transcript?: string };
  },
  options?: Partial<OpenAI.Chat.ChatCompletionCreateParams>,
  existingContentDeltas?: InternalContentDelta[],
): ContentDelta[] {
  const contentDeltas: ContentDelta[] = [];

  // OpenAI will only either have text, audio, or tool_calls
  // so we can assume that the first non-tool call index is 0
  const nonToolCallIndex = 0;

  // However, it has been noticed that OpenAI will sometimes
  // responds with a tool call after responding with text parts
  // therefore, we must increase the indexes of tool calls by this
  // TODO: we need to find a better solution here to map the indexes
  // correctly
  const nonToolCallCount =
    existingContentDeltas?.filter((delta) => delta.part.type !== "tool-call")
      .length || 0;

  if (delta.content) {
    contentDeltas.push({
      index: nonToolCallIndex,
      part: { type: "text", text: delta.content },
    });
  }
  if (delta.audio) {
    contentDeltas.push({
      index: nonToolCallIndex,
      part: {
        type: "audio",
        ...(delta.audio.data && {
          audioData: delta.audio.data,
          ...mapOpenAIAudioFormat(options?.audio?.format || "pcm16"),
          sampleRate: OPENAI_AUDIO_SAMPLE_RATE,
          channels: OPENAI_AUDIO_CHANNELS,
        }),
        ...(delta.audio.transcript && { transcript: delta.audio.transcript }),
      },
    });
  }
  if (delta.tool_calls) {
    for (const toolCall of delta.tool_calls) {
      contentDeltas.push({
        index: nonToolCallCount + toolCall.index,
        part: {
          type: "tool-call",
          ...(toolCall.id && { toolCallId: toolCall.id }),
          ...(toolCall.function?.name && { toolName: toolCall.function.name }),
          ...(toolCall.function?.arguments && {
            args: toolCall.function.arguments,
          }),
        },
      });
    }
  }
  return contentDeltas;
}

export function mapOpenAIUsage(usage: OpenAI.CompletionUsage): ModelUsage {
  return {
    inputTokens: usage.prompt_tokens,
    outputTokens: usage.completion_tokens,
    ...(usage.prompt_tokens_details && {
      inputTokensDetail: mapOpenAIPromptTokensDetails(
        usage.prompt_tokens_details as OpenAIPatchedPromptTokensDetails,
      ),
      outputTokensDetail: mapOpenAICompletionTokenDetails(
        usage.completion_tokens_details as OpenAIPatchedCompletionTokenDetails,
      ),
    }),
  };
}

export function mapOpenAIPromptTokensDetails(
  details: OpenAIPatchedPromptTokensDetails,
): ModelTokensDetail {
  return {
    textTokens: details.text_tokens,
    imageTokens: details.image_tokens,
    audioTokens: details.audio_tokens,
  };
}

export function mapOpenAICompletionTokenDetails(
  details: OpenAIPatchedCompletionTokenDetails,
): ModelTokensDetail {
  return {
    textTokens: details.text_tokens,
    audioTokens: details.audio_tokens,
    // note: reasoning_tokens is included in output_tokens
  };
}

File: sdk-rust/src/openai/openai.rs
use crate::{
    errors::{LanguageModelError, LanguageModelResult},
    language_model::LanguageModel,
    types::{
        AudioFormat, AudioPart, LanguageModelInput, Message, Modality, ModelResponse, Part,
        ResponseFormatOption, TextPart, Tool, ToolCallPart, ToolChoiceOption,
    },
};
use reqwest::{
    header::{self, HeaderValue},
    Client,
};

use super::openai_api;

const OPENAI_AUDIO_SAMPLE_RATE: i64 = 24_000;
const OPENAI_AUDIO_CHANNELS: i64 = 1;

pub struct OpenAIModel {
    pub model_id: String,
    pub base_url: String,
    pub client: Client,
    pub structured_outputs: bool,
}

pub struct OpenAIModelOptions {
    pub base_url: Option<String>,
    pub model_id: String,
    pub api_key: String,
    pub structured_outputs: bool,
}

impl OpenAIModel {
    pub fn new(options: OpenAIModelOptions) -> OpenAIModel {
        let mut headers = header::HeaderMap::new();
        let mut auth_header_value: HeaderValue =
            format!("Bearer {}", options.api_key).try_into().unwrap();
        auth_header_value.set_sensitive(true);
        headers.insert(header::AUTHORIZATION, auth_header_value);
        headers.insert(
            header::CONTENT_TYPE,
            HeaderValue::from_static("application/json"),
        );

        OpenAIModel {
            model_id: options.model_id,
            base_url: options
                .base_url
                .unwrap_or("https://api.openai.com/v1".to_string()),
            client: Client::builder().default_headers(headers).build().unwrap(),
            structured_outputs: options.structured_outputs,
        }
    }
}

#[async_trait::async_trait]
impl LanguageModel for OpenAIModel {
    fn provider(&self) -> &'static str {
        "openai"
    }

    fn model_id(&self) -> String {
        self.model_id.to_string()
    }

    async fn generate(&self, input: LanguageModelInput) -> LanguageModelResult<ModelResponse> {
        let params =
            convert_to_openai_params(input, self.model_id.clone(), self.structured_outputs)?;

        let response = self
            .client
            .post(format!("{}/chat/completions", self.base_url))
            .json(&params)
            .send()
            .await?;

        if response.status().is_client_error() {
            return Err(LanguageModelError::ClientError(response.text().await?));
        }

        let json = response.json::<openai_api::ChatCompletion>().await?;

        let choice = json.choices.first().ok_or(LanguageModelError::Invariant(
            "no choices in response".to_string(),
        ))?;

        let openai_api::ChatCompletionChoice { message, .. } = choice;

        if let Some(refusal) = &message.refusal {
            return Err(LanguageModelError::Refusal(refusal.to_string()));
        }

        let content = map_openai_message(message, params.audio.map(|audio| audio.format))?;

        Ok(ModelResponse {
            content,
            cost: None,
            usage: None,
        })
    }
}

fn convert_to_openai_params(
    input: LanguageModelInput,
    model_id: String,
    structured_outputs: bool,
) -> LanguageModelResult<openai_api::ChatCompletionCreateParamsBase> {
    Ok(openai_api::ChatCompletionCreateParamsBase {
        model: model_id,
        messages: convert_to_openai_messages(&input)?,
        max_tokens: input.max_tokens,
        temperature: input.temperature,
        top_p: input.top_p,
        presence_penalty: input.presence_penalty,
        frequency_penalty: input.frequency_penalty,
        seed: input.seed,
        tools: input.tools.as_ref().map(|tools| {
            tools
                .iter()
                .map(|tool| convert_to_openai_tool(tool, structured_outputs))
                .collect()
        }),
        tool_choice: input
            .tool_choice
            .as_ref()
            .map(convert_to_openai_tool_choice),
        response_format: input.response_format.as_ref().map(|response_format| {
            convert_to_openai_response_format(response_format, structured_outputs)
        }),
        modalities: input
            .modalities
            .map(|modalities| modalities.iter().map(convert_to_openai_modality).collect()),
        audio: input
            .extra
            .as_ref()
            .and_then(|extra| extra.get("audio"))
            .and_then(|value| {
                serde_json::from_value::<openai_api::ChatCompletionAudioParam>(value.clone()).ok()
            }),
        extra: input.extra,
    })
}

fn convert_to_openai_messages(
    input: &LanguageModelInput,
) -> LanguageModelResult<Vec<openai_api::ChatCompletionMessageParam>> {
    let mut openai_messages = vec![];

    if let Some(system_prompt) = &input.system_prompt {
        openai_messages.push(openai_api::ChatCompletionMessageParam::System(
            openai_api::ChatCompletionSystemMessageParam {
                content: system_prompt.to_string(),
            },
        ));
    }

    for message in &input.messages {
        match message {
            Message::Assistant(message) => {
                let mut openai_message_param = openai_api::ChatCompletionAssistantMessageParam {
                    content: None,
                    audio: None,
                    tool_calls: None,
                };

                for content in &message.content {
                    match content {
                        Part::Text(part) => {
                            openai_message_param
                                .content
                                .get_or_insert_with(Vec::new)
                                .push(openai_api::ChatCompletionContentPartText {
                                    text: part.text.clone(),
                                });
                        }
                        Part::ToolCall(part) => {
                            openai_message_param
                                .tool_calls
                                .get_or_insert_with(Vec::new)
                                .push(openai_api::ChatCompletionMessageToolCall {
                                    id: part.tool_call_id.clone(),
                                    function: openai_api::ChatCompletionMessageToolCallFunction {
                                        name: part.tool_name.clone(),
                                        arguments: part
                                            .args
                                            .as_ref()
                                            .map(|args| args.to_string())
                                            .unwrap_or_default(),
                                    },
                                    type_: "function".to_string(),
                                });
                        }
                        Part::Audio(part) => {
                            if let Some(id) = &part.id {
                                openai_message_param.audio =
                                    Some(openai_api::ChatCompletionAssistantMessageParamAudio {
                                        id: id.to_string(),
                                    });
                            } else {
                                return Err(LanguageModelError::InvalidInput(
                                    "audio part must have an id".to_string(),
                                ));
                            }
                        }
                        _ => Err(LanguageModelError::InvalidInput(
                            "unsupported content part for assistaant message".to_string(),
                        ))?,
                    }
                }

                openai_messages.push(openai_api::ChatCompletionMessageParam::Assistant(
                    openai_message_param,
                ));
            }
            Message::Tool(message) => {
                for content in &message.content {
                    openai_messages.push(openai_api::ChatCompletionMessageParam::Tool(
                        openai_api::ChatCompletionToolMessageParam {
                            tool_call_id: content.tool_call_id.clone(),
                            content: content.result.to_string(),
                        },
                    ));
                }
            }
            Message::User(message) => {
                let mut openai_message_param =
                    openai_api::ChatCompletionUserMessageParam { content: vec![] };

                for content in &message.content {
                    match content {
                        Part::Text(part) => {
                            openai_message_param.content.push(
                                openai_api::ChatCompletionContentPart::Text(
                                    openai_api::ChatCompletionContentPartText {
                                        text: part.text.clone(),
                                    },
                                ),
                            );
                        }
                        Part::Image(part) => {
                            openai_message_param.content.push(
                                openai_api::ChatCompletionContentPart::Image(
                                    openai_api::ChatCompletionContentPartImage {
                                        image_url: openai_api::ChatCompletionContentPartImageUrl {
                                            url: format!(
                                                "data:{};base64,{}",
                                                part.mime_type, part.image_data
                                            ),
                                            detail: None,
                                        },
                                    },
                                ),
                            );
                        }
                        Part::Audio(part) => {
                            openai_message_param.content.push(
                                openai_api::ChatCompletionContentPart::InputAudio(
                                    openai_api::ChatCompletionContentPartInputAudio {
                                        input_audio:
                                            openai_api::ChatCompletionContentPartInputAudioData {
                                                data: part.audio_data.clone(),
                                                format: match part.format {
                                                    None => Err(LanguageModelError::InvalidInput(
                                                        "audio part must have a format".to_string(),
                                                    )),
                                                    Some(AudioFormat::Wav) => Ok(openai_api::ChatCompletionContentPartInputAudioFormat::Wav),
                                                    Some(AudioFormat::Mp3) => Ok(openai_api::ChatCompletionContentPartInputAudioFormat::Mp3),
                                                    _ => Err(LanguageModelError::InvalidInput(
                                                        "unsupported audio format".to_string(),
                                                    )),
                                                }?,
                                            },
                                    },
                                ),
                            );
                        }
                        _ => {
                            return Err(LanguageModelError::InvalidInput(
                                "unsupported content part for user message".to_string(),
                            ))
                        }
                    }
                }

                openai_messages.push(openai_api::ChatCompletionMessageParam::User(
                    openai_message_param,
                ));
            }
        }
    }

    Ok(openai_messages)
}

fn convert_to_openai_tool(tool: &Tool, structured_outputs: bool) -> openai_api::ChatCompletionTool {
    openai_api::ChatCompletionTool {
        function: openai_api::ChatCompletionToolFunction {
            name: tool.name.clone(),
            description: Some(tool.description.clone()),
            parameters: tool.parameters.clone(),
            strict: Some(structured_outputs),
        },
        type_: "function".to_string(),
    }
}

fn convert_to_openai_tool_choice(
    tool_choice: &ToolChoiceOption,
) -> openai_api::ChatCompletionToolChoiceOption {
    match tool_choice {
        ToolChoiceOption::None => openai_api::ChatCompletionToolChoiceOption::None,
        ToolChoiceOption::Auto => openai_api::ChatCompletionToolChoiceOption::Auto,
        ToolChoiceOption::Required => openai_api::ChatCompletionToolChoiceOption::Required,
        ToolChoiceOption::Tool(tool) => openai_api::ChatCompletionToolChoiceOption::Named(
            openai_api::ChatCompletionNamedToolChoice {
                function: openai_api::ChatCompletionNamedToolChoiceFunction {
                    name: tool.tool_name.clone(),
                },
            },
        ),
    }
}

fn convert_to_openai_response_format(
    response_format: &ResponseFormatOption,
    structured_outputs: bool,
) -> openai_api::ResponseFormat {
    match response_format {
        ResponseFormatOption::Text => openai_api::ResponseFormat::Text,
        ResponseFormatOption::Json(format) => {
            if structured_outputs {
                if let Some(schema) = format.schema.as_ref() {
                    return openai_api::ResponseFormat::JSONSchema(
                        openai_api::ResponseFormatJSONSchema {
                            json_schema: openai_api::ResponseFormatJSONSchemaContent {
                                name: format.name.clone(),
                                strict: true,
                                schema: schema.clone(),
                            },
                        },
                    );
                }
            }
            openai_api::ResponseFormat::JSONObject
        }
    }
}

fn convert_to_openai_modality(modality: &Modality) -> openai_api::ChatCompletionModality {
    match modality {
        Modality::Text => openai_api::ChatCompletionModality::Text,
        Modality::Audio => openai_api::ChatCompletionModality::Audio,
    }
}

fn map_openai_message(
    message: &openai_api::ChatCompletionMessage,
    input_audio_format: Option<openai_api::ChatCompletionAudioParamFormat>,
) -> LanguageModelResult<Vec<Part>> {
    let mut parts: Vec<Part> = vec![];

    if let Some(content) = message.content.as_ref() {
        parts.push(Part::Text(TextPart {
            text: content.to_string(),
            id: None,
        }));
    }

    if let Some(completion_audio) = message.audio.as_ref() {
        let input_audio_format = {
            if let Some(format) = input_audio_format {
                format
            } else {
                return Err(LanguageModelError::Invariant(
                    "receive audio without input audio format".to_string(),
                ));
            }
        };

        let (channels, sample_rate) = match input_audio_format {
            openai_api::ChatCompletionAudioParamFormat::Pcm16 => {
                (Some(OPENAI_AUDIO_CHANNELS), Some(OPENAI_AUDIO_SAMPLE_RATE))
            }
            _ => (None, None),
        };

        let audio_part = AudioPart {
            id: Some(completion_audio.id.clone()),
            audio_data: completion_audio.data.clone(),
            // set to constant if input_audio_format = pcm, if not, None
            channels,
            sample_rate,
            transcript: Some(completion_audio.transcript.clone()),
            format: Some(match input_audio_format {
                openai_api::ChatCompletionAudioParamFormat::Pcm16 => AudioFormat::Linear16,
                openai_api::ChatCompletionAudioParamFormat::Mp3 => AudioFormat::Mp3,
                openai_api::ChatCompletionAudioParamFormat::Flac => AudioFormat::Flac,
                openai_api::ChatCompletionAudioParamFormat::Opus => AudioFormat::Opus,
                openai_api::ChatCompletionAudioParamFormat::Wav => AudioFormat::Wav,
            }),
        };

        parts.push(Part::Audio(audio_part));
    }

    if let Some(tool_calls) = message.tool_calls.as_ref() {
        for tool_call in tool_calls {
            parts.push(Part::ToolCall(ToolCallPart {
                tool_call_id: tool_call.id.clone(),
                tool_name: tool_call.function.name.clone(),
                args: serde_json::from_str(tool_call.function.arguments.as_str()).map_err(|e| {
                    LanguageModelError::InvalidInput(format!("invalid json: {}", e))
                })?,
                id: None,
            }))
        }
    }

    Ok(parts)
}
